cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/queue/JobQueueDataset.java

```diff
@@ -112,26 +109,55 @@ public void addNotification(ProgramSchedule schedule, Notification notification)
     try (CloseableIterator<Job> jobs = getJobsForSchedule(schedule.getScheduleId())) {
       while (jobs.hasNext()) {
         Job job = jobs.next();
-        if (job.getState() != Job.State.PENDING_LAUNCH) {
+        if (job.getState() == Job.State.PENDING_TRIGGER) {
           // check if its PENDING_LAUNCH. If so, create a new Job to avoid the chance that the job-launching
           // process has conflict
           addNotification(job, notification);
           existingJobModified = true;
+          break;
         }
       }
     }
     // if no existing job was modified with the new notification, add a new job with the first notification
     if (!existingJobModified) {
-      Job job = new SimpleJob(schedule, System.currentTimeMillis(),
-                              Lists.newArrayList(notification), Job.State.PENDING_TRIGGER);
-      put(job);
+      addNotification(new SimpleJob(schedule, System.currentTimeMillis(),
```

Comment:
Is the motivation for keeping Job immutable is for future in-memory cache? Otherwise, since we currently just store Jobs in the dataset, keeping the deserialized Job object immutable is not necessary.